tests/integration/results.go

```diff
@@ -16,29 +16,41 @@ import (
 	"testing"
 	"time"
 
+	"github.com/ipfs/go-cid"
 	"github.com/sourcenetwork/immutable"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/sourcenetwork/defradb/client"
 )
 
+// Validator instances can be substituted in place of concrete values
+// and will be asserted on using their [Validate] function instead of
+// asserting direct equality.
+//
+// They may mutate test state.
+type Validator interface {
+	Validate(s *state, actualValue any, msgAndArgs ...any)
+}
```

Comment:
praise: that's cool stuff. I like this approach very much as it will allow us easily introduce individual matchers instead of checking exact equality. 

I actually wanted myself to do something similar. I was even about to write a task a few weeks back, but something distracted me.

suggestion: when I was thinking about this the other day I had an ideal of using one of well established matching frameworks. My preference was [Gomega](https://onsi.github.io/gomega/#provided-matchers). There are other ones as well like [GoCrest](https://github.com/corbym/gocrest).

For example Gomega matcher has the following interface:
```go
type GomegaMatcher interface {
    Match(actual interface{}) (success bool, err error)
    FailureMessage(actual interface{}) (message string)
    NegatedFailureMessage(actual interface{}) (message string)
}
```
They have a bunch of very handy matchers for maps, slices, strings, async stuff and others.

Sadly [testify asserts](https://github.com/stretchr/testify/blob/master/assert/assertions.go) can't be used here, as they execute assertions right away and use no interfaces, hence can't be customized.

I do not propose to use them right now, because there a value in keeping things simple (as you did) for now. But it's something to consider (maybe even now). 

But I believe replicating their matcher interface makes a lot of sense.
That means matchers don't do asserts themselves (e.g. no `require.Equal`...) they just match and return `(bool, error)`. If it returns `false` we can ask it for the reason by calling `FailureMessage`. If we don't do this, we won't be able to combine our matchers (or validators).
For example, it works fine if you use a single matcher (like in your examples): 
```go
"cid": testUtils.NewUniqueCid("name"),
```
But what if you want to combine them like:
```go
"cid": testUtils.AnyOf(
    testUtils.NewUniqueCid("name"),
    testUtils.NewUniqueCid("age"),
)
```
There is no way you can do it, because `testUtils.NewUniqueCid("name")` will panic right away if it doesn't match.

Reply:
> But what if you want to combine them

That is a really nice thought - thanks Islam!  And I like the `GomegaMatcher` interface.

Like you seemed to hint, I also am a little torn as to whether to bring that complexity in now or not - I'll see what the others say but am leaning towards keeping it simple in the short term.  Changing this later should not require changes to the tests, so it should be easy enough to postpone it.

If leaving as is, I'll create an link a new issue via a code-comment todo so that we don't lose this thought.

- [ ] Implement or create and link ticket

Reply:
+1 for gomega, the Ω can seem a bit trippy at first haha.